### PR TITLE
Replace sharness-based test suite by pytest

### PR DESCRIPTION
**Pull request by ejoerns**
In the past, we faced many limitations of the sharness-based RAUC test suite and additionally, the growing size of the `rauc.t` file made it harder to keep oversight over the test cases.

With the sharness-based tests, we were often limited to having a simple program exit code check. But the reasons for exiting in RAUC are manifold.
With pytest, we can easily check for expected stdout/stderr output. Almost all tests were extended with this simple enhancement.

The new tests are now also split up into different test files, based on their test purpose.

### CODE REVIEW

**Review by jluebbe — COMMENTED** [review]
Instead of replicating the `skipif` marks everywhere, they could be defined once in a helper module as suggested here: https://docs.pytest.org/en/8.2.x/how-to/skipping.html#id1

You should probably run https://github.com/astral-sh/ruff on the new code. I'd suggest a [.ruff.toml](https://docs.astral.sh/ruff/configuration/#__tabbed_1_2) with `line-length = 119` (the same our other projects).

File: test/meson.build
Comment (ejoerns) [reply test/meson.build:72]: `test_cmdline` is a generic test for some comment line things while the other test were meant to test the subcommand they are named after. However, this is not fully true since I have tests named `test_rauc_sign` or `test_rauc_verify`, too. Suggestions welcome.

File: test/conftest.py
Comment (ejoerns) [reply test/conftest.py:265]: Indeed. This underwent a number of changes until it ended up like this. In the current setup a fixture is what it should be. fixed.

File: test/conftest.py
Comment (ejoerns) [reply test/conftest.py:18]: I was lazy here at first. The `monkeypatch` fixture used has a function scope and mixing scopes is not possible

**Comment by ejoerns** [comment]
@jluebbe Thank you for the quick review!

> Instead of replicating the `skipif` marks everywhere, they could be defined once in a helper module as suggested here: https://docs.pytest.org/en/8.2.x/how-to/skipping.html#id1

Indeed. While the skipif statements were easy to copy&paste, they are probably not that easy to maintain and change and pollute the set files a bit. Fixed.

> You should probably run https://github.com/astral-sh/ruff on the new code. I'd suggest a [.ruff.toml](https://docs.astral.sh/ruff/configuration/#__tabbed_1_2) with `line-length = 119` (the same our other projects).

Ran `ruff` on it and added `ruff` to our style-check flow.

Also the other review comments should be addressed by this updated version.

File: test/meson.build
Comment (jluebbe) [reply test/meson.build:72]: I'd just drop the `_rauc` part, as they all apply to RAUC.

File: test/conftest.py
Comment (jluebbe) [reply test/conftest.py:18]: https://github.com/pytest-dev/pytest/issues/363#issuecomment-1335631998 has a good workaround.

File: test/test_cmdline.py
Comment (jluebbe) [reply test/test_cmdline.py]: `test_rauc_excess_args` and `test_rauc_help` could use this approach as well.

**Comment by jluebbe** [comment]
As we don't have pytest in our container, this the tests are not yet running in CI. We should add the necessary dependencies in a separate PR first.

File: test/conftest.py
Comment (jluebbe) [reply test/conftest.py]: Do you want to keep pydbus for other tests that would be added later? If it's just for checking if the service is started, I'd prefer to avoid the dependency.

File: test/.ruff.toml
Comment (jluebbe) [inline test/.ruff.toml:3]: I've created https://github.com/rauc/rauc/pull/1406 to avoid this exclude.

File: test/conftest.py
Comment (ejoerns) [reply test/conftest.py:18]: Yes, I came across this or a similar, too. I'll give it a try then..

File: test/conftest.py
Comment (ejoerns) [reply test/conftest.py]: I've kept it first of all and have a prepared example on how to use it for dbus tests: https://github.com/ejoerns/rauc/commit/aa33a7ef08a34782c523d281898d3bf8270c5ca1

File: test/meson.build
Comment (ejoerns) [reply test/meson.build:92]: Right, I wanted to add a note about this. I've added it to the commit message.

**Comment by jluebbe** [comment]
Do you also get the splice warnings (`WARNING: CPU: 1 PID: 6623 at lib/iov_iter.c:1201 iov_iter_pipe+0x31/0x40`) when running locally?

**Comment by ejoerns** [comment]
> Do you also get the splice warnings (`WARNING: CPU: 1 PID: 6623 at lib/iov_iter.c:1201 iov_iter_pipe+0x31/0x40`) when running locally?

Yes, seems to be 9pfs-related.

**Comment by ejoerns** [comment]
Added some missing bundle copying in convert tests that previously failed for cross (386) only (for some reason).

File: test/meson.build
Comment (jluebbe) [reply test/meson.build:92]: It looks like with pytest 7.3, the number of tmpdirs to keep becomes configurable:
https://docs.pytest.org/en/7.3.x/reference/reference.html#confval-tmp_path_retention_count

This is available in Debian testing, so might land in backports at some point.

**Comment by ejoerns** [comment]
Applied two minor cleanups

**Comment by jluebbe** [comment]
Force-pushed to sort imports and to remove some redundant comments.

File: test/test_verify.py
Comment (jluebbe) [inline test/test_verify.py:23]: This should also be a module level mark.

File: test/test_replace_signature.py
Comment (jluebbe) [inline test/test_replace_signature.py:33]: The exitcode check is missing here and in the tests below..

File: test/test_replace_signature.py
Comment (jluebbe) [inline test/test_replace_signature.py:240]: Missing `t` at the end of the function name.

**Review by jluebbe — CHANGES_REQUESTED** [review]
Only a few minor things left. Other improvements can wait until this is merged.

File: test/test_install.py
Comment (ejoerns) [reply test/test_install.py]: Moved the test case to `test_cmdline.py` now.

**Review by jluebbe — APPROVED** [review]

**Pull request merged**
